pyro/distributions/multivariate_normal.py

```diff
@@ -0,0 +1,123 @@
+        if not sigma.dim() == 2:
+            raise ValueError("The covariance matrix must be a matrix, but got sigma.size() = {}".format(mu.size()))
+
+        super(MultivariateNormal, self).__init__(*args, **kwargs)
+
+    def batch_shape(self, x=None):
+        """
+        Ref: :py:meth:`pyro.distributions.distribution.Distribution.batch_shape`
+        """
+        mu = self.mu.expand(self.batch_size, *self.mu.size()).squeeze(0)
+        if x is not None:
+            if x.size()[-1] != mu.size()[-1]:
+                raise ValueError("The event size for the data and distribution parameters must match.\n"
+                                 "Expected x.size()[-1] == self.mu.size()[0], but got {} vs {}".format(
+                                    x.size(-1), mu.size(-1)))
+            try:
+                mu = mu.expand_as(x)
+            except RuntimeError as e:
+                raise ValueError("Parameter `mu` with shape {} is not broadcastable to "
+                                 "the data shape {}. \nError: {}".format(mu.size(), x.size(), str(e)))
+
+        return mu.size()[:-1]
+
+    def event_shape(self):
+        """
+        Ref: :py:meth:`pyro.distributions.distribution.Distribution.event_shape`
+        """
+        return self.mu.size()[-1:]
+
+    def sample(self):
+        """
+        Generate a sample with the specified covariance matrix and mean.
+        Differentiation wrt. to Sigma is only supported on PyTorch version 0.3.0 or higher.
+        Ref: :py:meth:`pyro.distributions.distribution.Distribution.sample`
+
+        """
+        batch_size = self.batch_size
+        uncorrelated_standard_sample = Variable(torch.randn(batch_size, *self.mu.size()).type_as(self.mu.data))
+        transformed_sample = self.mu + torch.mm(uncorrelated_standard_sample, self.sigma_cholesky)
+        return transformed_sample if self.reparameterized else transformed_sample.detach()
+
+    def batch_log_pdf(self, x, normalized=True):
+        """
+        Return the logarithm of the probability density function evaluated at x.
+
+        :param x: The points for which the log_pdf should be evaluated batched along axis 0.
+        :param normalized: If set to `False` the normalization constant is omitted in the results.
+            This might be preferable, as computing the determinant of sigma might not always be numerically stable.
+            Defaults to `True`.
+        :return: A `torch.autograd.Variable` of size `self.batch_shape(x) + (1,)`
+
+        Ref: :py:meth:`pyro.distributions.distribution.Distribution.batch_log_pdf`
+        """
+        batch_size = x.size()[0] if len(x.size()) > len(self.mu.size()) else 1
+        batch_log_pdf_shape = self.batch_shape(x) + (1,)
+        x = x.view(batch_size, *self.mu.size())
+        normalization_factor = torch.log(
+            self.sigma_cholesky.diag()).sum() + (self.mu.size()[0] / 2) * np.log(2 * np.pi) if normalized else 0
```

Comment:
I guess the gradients will be incorrect if `sigma_cholesky.requires_grad=True` and `normalized=False`. Do you think it would be reasonable to either (1) raise a `NotImplementedError` in this case (to ensure users set `sigma_cholesky.requires_grad=True`) or (2) compute a numerically-stable quantity whose value is zero but whose gradient agrees with the gradient of the normalization factor (this is a research question 😉 )?

Reply:
I see the problem there, of course the gradient of the normalization coefficient is not taken into account at the moment. I am however not completely sure how much of a problem this is, as I would presume that the normalization might not be of real importance for many applications. I must admit that I have not thought too much about this problem until now, but I will look into it. Approach (2) seems better to me though, as (1) would disable differentiation for those who don't care about normalization too.

Reply:
If (2) is difficult, I think it's fine to merge (1) for now and implement (2) later once we know how. I'd like to merge your PR soon since @fehiepsi is already building a Gaussian Process tutorial #650 and it would be nice if he could base that on your work.

Reply:
Regarding (1) I have added a warning about the incorrect gradients to `batch_log_pdf`. This seems to be the best option to me, as not being able to differentiate if normalized is set to false would seriously limit the usefulness of this feature in my opinion.

Reply:
Approach (2) might also not be as difficult as it seems. The gradient of the determinant of X apparently is ` 2*X.inverse() - torch.diag(torch.diag(X.inverse()))`, a rather nice expression (see (141) in http://www.math.uwaterloo.ca/~hwolkowi//matrixcookbook.pdf) especially since the inverse is already being computed. I am however not sure how to inject that gradient into autograd, do you have any suggestions as to how that might be achieved efficiently?

Reply:
I think the easiest way is to create an autograd function:
```py
from torch.autograd import Function
class _NonnormalizedDeterminate(Function):
    @staticmethod
    def forward(ctx, matrix):
        ctx.save_for_backward(matrix)
        return matrix.new([1.0])

    @staticmethod
    @once_differentiable
    def backward(ctx, grad_output):
        matrix, = ctx.saved_tensors
        inv = matrix.inverse()
        grad = 2 * inv - torch.diag(torch.diag(inv))
        return grad_output * grad
```
You can then use this as
```py
normalization_factor = torch.log(_NonnormalizedDeterminant.apply(sef.sigma)) + ...
```

Reply:
Also feel free to open an issue on this so we don't lose this extended discussion.